### PR TITLE
across()

### PR DESCRIPTION
**Pull request by romainfrancois**
``` r
library(dplyr, warn.conflicts = FALSE)

iris %>%
  group_by(Species) %>%
  summarise(across(starts_with("Sepal"), mean))
#> # A tibble: 3 x 3
#>   Species    Sepal.Length Sepal.Width
#>   <fct>             <dbl>       <dbl>
#> 1 setosa             5.01        3.43
#> 2 versicolor         5.94        2.77
#> 3 virginica          6.59        2.97

iris %>%
  group_by(Species) %>%
  summarise(across(starts_with("Sepal"), ~mean(.x, na.rm = TRUE)))
#> # A tibble: 3 x 3
#>   Species    Sepal.Length Sepal.Width
#>   <fct>             <dbl>       <dbl>
#> 1 setosa             5.01        3.43
#> 2 versicolor         5.94        2.77
#> 3 virginica          6.59        2.97

# a named list of functions
iris %>%
  group_by(Species) %>%
  summarise(across(starts_with("Sepal"), list(mean = mean, sd = sd)))
#> # A tibble: 3 x 3
#>   Species    mean$Sepal.Length $Sepal.Width sd$Sepal.Length $Sepal.Width
#>   <fct>                  <dbl>        <dbl>           <dbl>        <dbl>
#> 1 setosa                  5.01         3.43           0.352        0.379
#> 2 versicolor              5.94         2.77           0.516        0.314
#> 3 virginica               6.59         2.97           0.636        0.322
```

<sup>Created on 2019-11-28 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

### CODE REVIEW

**Comment by romainfrancois** [comment]
I've added the function `current_key()` we can probably come up with a better name so that we can query the keys for the current group, i.e. this is the same as `.y` in `group_modify()` 

This complements `pick()` which gives the current slice. 

We can't just use `pick()` because of potentially empty groups.

**Comment by romainfrancois** [comment]
closes #4540

**Comment by romainfrancois** [comment]
closes #4517

**Comment by romainfrancois** [comment]
closes #3769

**Comment by BenCarlsen** [comment]
I feel the name `across()` can be potentially confusing since the name suggests movement, progress or direction; in applications like `https://github.com/tidyverse/dplyr/issues/4517` `across()` is employed to _suppress_ the modify-the-source-after-each-argument behavior of dplyr verbs. 

We will be using a function whose name suggests an action with a meaningful direction to modify the behavior of an operation that has with a meaningful order (e.g. `mutate()`) so that it instead performs that operation in a manner in which order/progress is not meaningful.

**Comment by romainfrancois** [comment]
What do you suggest @BenCarlsen ?

**Comment by romainfrancois** [comment]
I've updated the code in the description of this issue to better reflect the current situation, and removed `pick()` as we can "just" use `across()` with `funs = identity`. 

Two arguments remaining: 
 - `select`: A tidy selection. For now, only equivalent to the old `_at()` but soon enough I'll incorporate changes to reflect the new abilities of `tidyselect`
 - `funs`: either a single function (or lambda) or a named list of functions. 

When we give a single functions, the output columns are named after the input columns. 

When giving a named list of functions, we get a pack for each function. 

Some points we need to discuss: 

 - The `current_column()` function gives the name of the column currently dealt with. I'm not a fan of the name, that can change. 

 - Should the picking of the involved column be done using :

    - the current slice: 
    - all the data

**Comment by lionel-** [comment]
Maybe the shorter `this_col()` and `this_key()`?

Also `funs` should be `fns`.

**Comment by lionel-** [comment]
> Should the picking of the involved column be done using

Maybe this could be parameterised? The default should probably be the whole data so `across()` and `select()` are consistent.

**Comment by flying-sheep** [comment]
If `current_column()` works in any `mutate_if` and so on, this also fixes #4511

I like the `.var` pronoun better than a function, as `.data` is also a pronoun.

**Comment by romainfrancois** [comment]
@lionel- I think at first I'll go with `eval_select()` in the full data, this way we consistently get the same variables for all groups.

**Comment by romainfrancois** [comment]
``` r
library(dplyr, warn.conflicts = FALSE)

iris %>%
  group_by(Species) %>%
  summarise(across(is.numeric, mean))
#> # A tibble: 3 x 5
#>   Species    Sepal.Length Sepal.Width Petal.Length Petal.Width
#>   <fct>             <dbl>       <dbl>        <dbl>       <dbl>
#> 1 setosa             5.01        3.43         1.46       0.246
#> 2 versicolor         5.94        2.77         4.26       1.33 
#> 3 virginica          6.59        2.97         5.55       2.03

iris %>%
  group_by(Species) %>%
  summarise(across(is.numeric, list(mean = mean, sd = sd)))
#> # A tibble: 3 x 3
#>   Species mean$Sepal.Leng… $Sepal.Width $Petal.Length $Petal.Width
#>   <fct>              <dbl>        <dbl>         <dbl>        <dbl>
#> 1 setosa              5.01         3.43          1.46        0.246
#> 2 versic…             5.94         2.77          4.26        1.33 
#> 3 virgin…             6.59         2.97          5.55        2.03 
#> # … with 4 more variables: sd$Sepal.Length <dbl>, $Sepal.Width <dbl>,
#> #   $Petal.Length <dbl>, $Petal.Width <dbl>
```

<sup>Created on 2019-11-29 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

@lionel- should lambda form work too, i.e.: 

``` r
library(dplyr, warn.conflicts = FALSE)
library(tidyselect)

iris %>%
  group_by(Species) %>%
  summarise(across(~is.numeric(.), mean))
#> Error: Must select with column names or positions.
#> ✖ `i` has the wrong type `logical`.
#> ℹ These indices must be positions or names.

eval_select(~is.numeric(.), data = iris)
#> Error: object '.' not found
eval_select(rlang::as_function(~is.numeric(.)), data = iris)
#> Sepal.Length  Sepal.Width Petal.Length  Petal.Width 
#>            1            2            3            4
```

<sup>Created on 2019-11-29 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

**Comment by lionel-** [comment]
> should lambda form work too, i.e.:

Yes I think so.

**Comment by flying-sheep** [comment]
Are the `_if`, `_at`, `_all` suffixes to be deprecated because of this?

Because if not, `current_column` should work in those, too.

**Comment by lionel-** [comment]
> Are the _if, _at, _all suffixes to be deprecated because of this?
>
> Because if not, current_column should work in those, too.

They are probably not going to be deprecated, but retired instead. Meaning we'll keep them in the package for backward compatibility as long as they don't take too much work to maintain. This does not imply we'll continue working on them and adding features.

**Comment by flying-sheep** [comment]
Then it makes sense. Should the text about retiring them be added in this PR?

File: R/each.R [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    fns <- as_function(fns)
+  } else {
+    if (is.null(names(fns))) {
+      abort("funs should be a single function, a single formula, or a named list of functions or formulas")
+    }
+    fns <- map(fns, as_function)
+  }
+
+  f <- function(df) {
+    if (single_function) {
```
Comment (hadley) [inline R/each.R]: Maybe move outside the `f()`?

File: R/each.R [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+#' @rdname across
+#' @export
+current_column <- function() {
+  context_env[["..current_column_name"]] %||% abort("No current column name registered, current_column() only makes sense inside across()")
+}
+
+#' Apply a function or a set of functions to a set of columns
+#'
+#' Creates a data frame by applying a set of functions to a tidy
+#' selection of columns in the current slice
```
Comment (hadley) [inline R/each.R]: This allows you to use `select()` semantics inside in `summarise()` and `mutate()`

File: R/each.R [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+#'  selected columns.
+#'
+#'  `current_column()` gives the name of the column currently being processed.
+#'
+#' @examples
+#'
+#' # A single function
+#' iris %>%
+#'   group_by(Species) %>%
+#'   summarise(across(starts_with("Sepal"), mean))
```
Comment (hadley) [inline R/each.R]: Show `rowMeans()` example too?

File: R/each.R [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+#' selection of columns in the current slice
+#'
+#' @param select tidy selection of columns
+#' @param fns Functions to apply to each of the selected columns. Possible
+#'   values are:
+#'
+#'   - A single function or a single quosure style lambda, e.g. `~ mean(.x, na.rm = TRUE)`
+#'   - A named list of functions and/or lambdas
+#'
+#'  @return A tibble
```
Comment (hadley) [inline R/each.R]: This and subsequent lines are indented, but shouldn't be

File: R/each.R [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  context_env[["..current_column_name"]] <- name
+}
+
+poke_current_column <- function(name) {
+  old <- context_env[["..current_column_name"]]
+  set_current_column(name)
+  old
+}
+
+#' @rdname across
```
Comment (hadley) [inline R/each.R]: I think it would be slightly better to document in `current_key()`

File: R/each.R [hunk 6]
```diff
@@ -0,0 +1,106 @@
+#' key data frame for the current group
```
Comment (hadley) [inline R/each.R]:
```suggestion
#' Metadata about the "current" element
```

File: R/each.R (hunk 1 above)
Comment (romainfrancois) [reply R/each.R]: We actually don't need `f` at all, it was a memory of previous implementation attempts

File: R/each.R (hunk 3 above)
Comment (romainfrancois) [reply R/each.R]: Deferring this to after the https://github.com/tidyverse/dplyr/issues/4673 conversation

**Pull request merged**